contents/javascript/javascript.md

```diff
@@ -0,0 +1,272 @@
+    var bodyColor = bodyColor || 'default';
+    ...
+}
+
+// es6 syntax default parameters
+function createPopUp(title, content, status, headerColor = 'default', bodyColor = 'default') {
+    ...
+}
+```
+
+ This is relatively simple and easy to understand. However, in order to specify body color, the user would have to know and fill in the default header color.
+
+```js
+// es5 syntax
+function createPopUp(title, content, status, optionals) {
+    var headerColor = optionals.headerColor || 'default';
+    var bodyColor = optionals.bodyColor || 'default';
+    ...
+}
+// es6 syntax with destructuring and default parameters
+function createPopUp(title, content, status, optionals) {
+    const { headerColor = 'default', bodyColor = 'default' } = optionals;
+    ...
+}
+```
+
+By using an object, the user would just need to fill in what parameter they want to be changed. In large functions, this would result in better readability as well.
+
+```js
+createPopUp('Warning', 'This will delete everything!', dangerStatus, { bodyColor: 'red' });
+```
+
+### Understand method chaining
+
+Method chaining is syntax such as `array.concat([1, 2]).push(1).filter(isEven)`. It is also sometimes referred to as the fluent interface.
+
+It is achieved by returning the object itself in the call.
+```js
+var obj = {
+    save: function() {
+        // some procedures
+        return this;
+    }
+};
+```
+
+Method chaining can be found in other languages but it is very common in Javascript. It can be found in libraries such as JQuery and lodash. Compare the two snippets below:
+
+```js
+var bob = new Person();
+
+bob.setName('Bob');
+bob.setAge(16);
+bob.setGender('male');
+
+bob.save();
+```
+
+```js
+new Person()
+  .setName('Bob')
+  .setAge(16)
+  .setGender('male')
+  .save();
+```
+
+Not only is the method chaining much shorter, it also makes the code more maintainable by keeping properties organised nicely and easily traceable in version control.
+
+However, like any design pattern, this is not to be abused. Only use method chaining when methods are related to the object at hand.
+
+### Don't reinvent the wheel
+
+The limitation to Javascript is its limited utility library. That is why libraries like [lodash](https://lodash.com/) is such a popular module. There are a lot of functions being provided, which would help a lot in making shorter, more coherent syntax. Lodash reduces the need to write and maintain your own `_.assign` or `_.range` and even provides powerful functions such as `_.merge` and `_.groupBy`.
+
+As a beginner, there may be some hesistation to use lodash or external utility libraries. The big advantage is the amount of benefit it brings to the team. Whether its the overhead of having to load the `lodash` library, or having to learn a large library's syntax, not having to maintain your own utility functions makes the project more manageable, maintainable and that alone makes it worth its cost.
+
+## Testing Libraries
```

Comment:
If there's going to be a document on testing for web apps, there's [Karma](https://github.com/karma-runner/karma) that launches multiple browsers automatically for testing (something similar to what the automated GUI testing does in Address Book)